service/worker/failoverManager/workflow.go

```diff
@@ -0,0 +1,376 @@
+	var totalNumOfDomains int
+	wfState := wfRunning
+	err = workflow.SetQueryHandler(ctx, QueryType, func(input []byte) (*QueryResult, error) {
+		return &QueryResult{
+			TotalDomains:   totalNumOfDomains,
+			Success:        len(successDomains),
+			Failed:         len(failedDomains),
+			State:          wfState,
+			TargetCluster:  params.TargetCluster,
+			SourceCluster:  params.SourceCluster,
+			SuccessDomains: successDomains,
+			FailedDomains:  failedDomains,
+		}, nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// get target domains
+	ao := workflow.WithActivityOptions(ctx, getGetDomainsActivityOptions())
+	getDomainsParams := &GetDomainsActivityParams{
+		TargetCluster: params.TargetCluster,
+		SourceCluster: params.SourceCluster,
+		Domains:       params.Domains,
+	}
+	var domains []string
+	err = workflow.ExecuteActivity(ao, GetDomainsActivity, getDomainsParams).Get(ctx, &domains)
+	if err != nil {
+		return nil, err
+	}
+	totalNumOfDomains = len(domains)
+
+	// failover in batch
+	ao = workflow.WithActivityOptions(ctx, getFailoverActivityOptions())
+	batchSize := params.BatchFailoverSize
+	times := len(domains) / batchSize
+
+	pauseCh := workflow.GetSignalChannel(ctx, PauseSignal)
+	resumeCh := workflow.GetSignalChannel(ctx, ResumeSignal)
+	var shouldPause bool
+
+	for i := 0; i < times; i++ {
+		// check if need to pause
+		shouldPause = pauseCh.ReceiveAsync(nil)
+		if shouldPause {
+			wfState = wfPaused
+			resumeCh.Receive(ctx, nil)
+		}
+		wfState = wfRunning
+
+		failoverActivityParams := &FailoverActivityParams{
+			Domains:       domains[i*batchSize : min((i+1)*batchSize, totalNumOfDomains)],
+			TargetCluster: params.TargetCluster,
+		}
+		var actResult FailoverActivityResult
+		workflow.ExecuteActivity(ao, FailoverActivity, failoverActivityParams).Get(ctx, &actResult)
+		successDomains = append(successDomains, actResult.SuccessDomains...)
+		failedDomains = append(failedDomains, actResult.FailedDomains...)
+
+		workflow.Sleep(ctx, time.Duration(params.BatchFailoverWaitTimeInSeconds)*time.Second)
+	}
+
+	wfState = wfCompleted
+	return &FailoverResult{
+		SuccessDomains: successDomains,
+		FailedDomains:  failedDomains,
+	}, nil
+}
+
+func getGetDomainsActivityOptions() workflow.ActivityOptions {
+	return workflow.ActivityOptions{
+		ScheduleToStartTimeout: 10 * time.Second,
+		StartToCloseTimeout:    20 * time.Second,
+		RetryPolicy: &cadence.RetryPolicy{
+			InitialInterval:    2 * time.Second,
+			BackoffCoefficient: 2,
+			MaximumInterval:    1 * time.Minute,
+			ExpirationInterval: 10 * time.Minute,
+			NonRetriableErrorReasons: []string{
+				errMsgParamsIsNil,
+				errMsgTargetClusterIsEmpty,
+				errMsgSourceClusterIsEmpty,
+				errMsgTargetClusterIsSameAsSource},
+		},
+	}
+}
+
+func getFailoverActivityOptions() workflow.ActivityOptions {
+	return workflow.ActivityOptions{
+		ScheduleToStartTimeout: 10 * time.Second,
+		StartToCloseTimeout:    10 * time.Second,
+	}
+}
+
+func min(a, b int) int {
```

Comment:
nit: there's already a function `MinInt()` defined in the common package.

Reply:
cool, removed